core/src/main/scala/kafka/server/KafkaRequestHandler.scala

```diff
@@ -69,10 +108,23 @@ class KafkaRequestHandler(id: Int,
           completeShutdown()
           return
 
+        case request: RequestChannel.CallbackRequest =>
+          try {
+            request.originalRequest.callbackRequestDequeTimeNanos = Some(time.nanoseconds())
```

Comment:
I wasn't sure if we would have a new request here though. We could be handling this callback after the original request returned right?

Reply:
Ah I misunderstood your comment. I guess I'm just not sure what setting current request here does for us. Why would we use currentRequest instead of original request? Unless we wanted to schedule another callback 😅

Reply:
> Unless we wanted to schedule another callback

We might in the future.  From the implementation perspective it may seem like we have "the request" and "the callback", but conceptually we should think in terms of a request that may run a few asynchronous operations and then continue processing, in which case we can have multiple callbacks in the context of the same request.  So logically it's like this:

-- request starts processing
1. Do some processing on a broker request thread
2. Do async operation (RPC, or maybe even async storage access in some distant future)
3. Continue processing on a broker request thread.
4. Do some async operation
5. Continue processing on a broker request thread
6. Do some async operation
7. Finish processing on a broker request thread

-- request is done

In this case the steps 1, 3, 5, 7 would be accounted as local time and logically we just keep processing the same request, only instead of blocking the thread in steps 2, 4, 6, we wait without blocking the thread.

Reply:
I think the issue is that the way that I implemented the metric is that we only have callback start and end. If we wanted to store multiple callbacks, we would need to change this. I'm tempted to tackle this sort of thing in a followup change to avoid overcomplicating this one.

Reply:
Yeah, if we need to make larger change to complete the framework, then we should definitely do in a follow-up change, but I think in this case all we need to do is this:
```
  val prevCallbacksTimeNanos = originalRequest.callbackRequestCompleteTimeNanos.getOrElse(0) - originalRequest.callbackRequestDequeTimeNanos.getOrElse(0)
  originalRequest.callbackRequestCompleteTimeNanos = None

  originalRequest.callbackRequestDequeTimeNanos = Some(time.nanoseconds() - prevCallbacksTimeNanos)
  currentRequest.set(callback.originalRequest)
```
and then we've got ourselves a fully functional framework for supporting arbitrary number of async calls in a request processing :-).